File: companies/ada_google_prep_program/Q1FindPatternsMatchingAnyWords.java
package ada_google_prep_program;

import java.util.*;

public class Q1FindPatternsMatchingAnyWords {
  private static boolean matches(String word, String pattern) {
    if (word.length() != pattern.length()) return false;
    for (int i = 0; i < pattern.length(); i++)
      if (word.charAt(i) != pattern.charAt(i) && pattern.charAt(i) != '.') return false;
    return true;
  }

  public static String[] findPatternsThatMatchAnyWords(String[] words, String[] patterns) {
    Collection<String> patternsThatMatchAnyWords = new ArrayList<>();
    for (String pattern : patterns)
      for (String word : words)
        if (matches(word, pattern)) {
          patternsThatMatchAnyWords.add(pattern);
          break;
        }
    return patternsThatMatchAnyWords.toArray(String[]::new);
  }

  public static void main(String[] args) {
    assert Arrays.equals(
      findPatternsThatMatchAnyWords(
        new String[]{"apple", "aws"}, new String[]{"a..le", "...", "....", "aws", "aws.", "aw"}
      ),
      new String[]{"a..le", "...", "aws"}
    );
  }
}

File: companies/ada_google_prep_program/q1_find_patterns_matching_any_words.go
package main

import "reflect"

func matches(word string, pattern string) bool {
	if len(word) != len(pattern) {
		return false
	}
	for i := 0; i < len(pattern); i++ {
		if word[i] != pattern[i] && pattern[i] != '.' {
			return false
		}
	}
	return true
}

func findPatternsThatMatchAnyWords(words []string, patterns []string) []string {
	patternsThatMatchAnyWords := make([]string, 0, len(patterns))
	for _, pattern := range patterns {
		for _, word := range words {
			if matches(word, pattern) {
				patternsThatMatchAnyWords = append(patternsThatMatchAnyWords, pattern)
				break
			}
		}
	}
	return patternsThatMatchAnyWords
}

func main() {
	if !reflect.DeepEqual(
		findPatternsThatMatchAnyWords([]string{"apple", "aws"}, []string{"a..le", "...", "....", "aws", "aws.", "aw"}),
		[]string{"a..le", "...", "aws"},
	) {
		panic("")
	}
}
